pipelined/bevy_sprite2/src/render/mod.rs

```diff
@@ -204,114 +240,84 @@ pub fn extract_sprites(
-struct SpriteVertex {
-    pub position: [f32; 3],
-    pub uv: [f32; 2],
+#[derive(Copy, Clone, Debug, Pod, Zeroable)]
+struct SpriteInstance {
+    pub transform: [[f32; 4]; 4],
+    pub sprite_size: [f32; 2],
+    pub uv_min: [f32; 2],
+    pub uv_size: [f32; 2],
 }
 
 pub struct SpriteMeta {
-    vertices: BufferVec<SpriteVertex>,
-    indices: BufferVec<u32>,
-    quad: Mesh,
+    /// The buffer containing sprite instance data
+    instances: BufferVec<SpriteInstance>,
+    /// The list of sprite batches
+    drawables: Vec<DrawableSpriteBatch>,
     view_bind_group: Option<BindGroup>,
-    texture_bind_group_keys: Vec<FrameSlabMapKey<Handle<Image>, BindGroup>>,
     texture_bind_groups: FrameSlabMap<Handle<Image>, BindGroup>,
 }
 
+/// A batch of sprites that share the same texture and can be drawn in the same call
+struct DrawableSpriteBatch {
+    texture_bind_group_key: FrameSlabMapKey<Handle<Image>, BindGroup>,
+    instances: Range<u32>,
+}
+
 impl Default for SpriteMeta {
     fn default() -> Self {
         Self {
-            vertices: BufferVec::new(BufferUsage::VERTEX),
-            indices: BufferVec::new(BufferUsage::INDEX),
+            instances: BufferVec::new(BufferUsage::VERTEX),
             texture_bind_groups: Default::default(),
-            texture_bind_group_keys: Default::default(),
+            drawables: Default::default(),
             view_bind_group: None,
-            quad: Quad {
-                size: Vec2::new(1.0, 1.0),
-                ..Default::default()
-            }
-            .into(),
         }
     }
 }
 
 pub fn prepare_sprites(
     render_device: Res<RenderDevice>,
     mut sprite_meta: ResMut<SpriteMeta>,
-    extracted_sprites: Res<ExtractedSprites>,
+    mut extracted_sprites: ResMut<ExtractedSprites>,
 ) {
     // dont create buffers when there are no sprites
     if extracted_sprites.sprites.is_empty() {
         return;
     }
 
-    let quad_vertex_positions = if let VertexAttributeValues::Float32x3(vertex_positions) =
-        sprite_meta
-            .quad
-            .attribute(Mesh::ATTRIBUTE_POSITION)
-            .unwrap()
-            .clone()
-    {
-        vertex_positions
-    } else {
-        panic!("expected vec3");
-    };
-
-    let quad_indices = if let Indices::U32(indices) = sprite_meta.quad.indices().unwrap() {
-        indices.clone()
-    } else {
-        panic!("expected u32 indices");
-    };
-
-    sprite_meta.vertices.reserve_and_clear(
-        extracted_sprites.sprites.len() * quad_vertex_positions.len(),
-        &render_device,
-    );
-    sprite_meta.indices.reserve_and_clear(
-        extracted_sprites.sprites.len() * quad_indices.len(),
-        &render_device,
-    );
-
-    for (i, extracted_sprite) in extracted_sprites.sprites.iter().enumerate() {
-        let sprite_rect = extracted_sprite.rect;
+    // Sort sprites by their image
+    extracted_sprites
+        .sprites
+        .sort_unstable_by(|a, b| a.handle.cmp(&b.handle));
```

Comment:
A `Ord` impl for the sprites type that stores this logic would allow us to just use `sort_unstable` here, and expose this natural ordering for others to reuse.

Reply:
Implementing `Ord` requires implementing `Eq`, but because of the `Vec2` types in `ExtractedSprite`, we can only implement `PartialEq`. Since `ExtractedSprite` should only be used internally I think this is fine, and I kind of like that it shows explicitly right here that we are sorting by the image handle.

Reply:
Sounds good to me.